### PR TITLE
Additions to base library

### PR DESCRIPTION
**Pull request by joehendrix**
This introduces a handful of additions to the base library.  The tuple, list, monad, and nat namespaces are affected.

### CODE REVIEW

File: library/data/list.lean [hunk 1]
```diff
@@ -41,6 +41,25 @@ theorem length_dropn
...
+    | (is_true  h₂) :=
+       is_true (calc a :: r = b :: r : congr_arg (λc, c :: r) h₁
+                        ... = b :: s : congr_arg (λt, b :: t) h₂)
+    | (is_false h₂) :=
+      is_false (λ q, list.no_confusion q (λ heq teq, h₂ teq))
+    end
+  | (is_false h₁) :=
+     is_false (λ q, list.no_confusion q (λ heq teq, h₁ heq))
+  end
+
```
Comment (leodemoura) [inline library/data/list.lean]: We already have an instance for list here:
https://github.com/leanprover/lean/blob/master/library/init/data/list/instances.lean#L25

File: library/data/tuple.lean [hunk 2]
```diff
@@ -45,6 +47,14 @@ theorem tail_cons (a : α) : Π (v : tuple α n), tail (a :: v) = v
...
 definition to_list : tuple α n → list α | ⟨ l, h ⟩ := l
 
+definition has_decidable_eq [decidable_eq α] {n:ℕ}
+  : ∀ (x y : tuple α n), decidable (x = y)
+| ⟨s,p⟩ ⟨t,q⟩ :=
+  match list.has_decidable_eq s t with
+  | (is_true h)  := is_true (subtype.eq h)
+  | (is_false h) := is_false (λr, subtype.no_confusion r (λleq (peq : p == q), h leq))
+  end
+
```
Comment (leodemoura) [inline library/data/tuple.lean]: Line 15 defines this instance.

File: library/data/tuple.lean [hunk 3]
```diff
@@ -119,3 +129,7 @@ section accum
 
 end accum
 end tuple
+
+instance decide_tuple_eq {A:Type.{1}} [decidable_eq A] {n:ℕ}
+  : ∀ {x y : tuple A n}, decidable (x = y)
+  := tuple.has_decidable_eq
```
Comment (leodemoura) [inline library/data/tuple.lean]: See line 15

File: library/init/category/combinators.lean [hunk 4]
```diff
@@ -47,6 +47,15 @@ def filter {m : Type → Type v} [monad m] {α : Type} (f : α → m bool) : lis
 | []       := return []
 | (h :: t) := do b ← f h, t' ← filter t, cond b (return (h :: t')) (return t')
 
+def foldl {m : Type u → Type v} [monad m] {s : Type u} {α : Type w} : (s → α → m s) → s → list α → m s
+| f s [] := return s
+| f s (h :: r) := do
+  s' ← f s h,
+  foldl f s' r
+
+def when {m : Type → Type} [monad m] (c : Prop) [h : decidable c] (t : m unit) : m unit :=
```
Comment (joehendrix) [inline library/init/category/combinators.lean:56]: I noticed there was a "when" def specific for tactic.  I haven't eliminated that but could.

File: library/data/list.lean (hunk 1 above)
Comment (joehendrix) [reply library/data/list.lean]: Thanks, I pushed a commit that removes the redundant decidable instances; I had these in a fork that had gotten stale.

**Comment by leodemoura** [comment]
@joehendrix Sorry for the delay. I will merge this PR after POPL.

**Pull request closed without merging**